Add implementation of complex numbers

sbastrakov opened the pull request:
This PR adds implementation of alpaka's own Complex class template. It is basically `std::complex<T>` but its methods also work on device. Math functions are (WIP now) done via alpaka traits, same way as for real numbers.

TODO:
- [x] Math traits. For now I've done only one for Abs, just to demonstrate how I would like to do it, and to discuss.
- [x] Unit tests.
- [x] Figure out what to do with SYCL math implementation.

Resolve #734.

sbastrakov commented on include/alpaka/math/abs/AbsStdLib.hpp:
```diff
@@ -29,17 +30,15 @@ namespace alpaka
-            struct Abs<
-                AbsStdLib,
-                TArg,
-                std::enable_if_t<std::is_arithmetic<TArg>::value && std::is_signed<TArg>::value>>
```
Note: these requirements did not work for `std::complex<T>` either, I think they were just wrong, but didn't make any difference before.

sbastrakov replied:
Ah, now after CI failed I understand the reason. After I removed that, some compilers issue a warning that `std::abs` call was removed as it is called for unsigned.

sbastrakov commented on include/alpaka/math/abs/AbsStdLib.hpp:
```diff
@@ -29,17 +30,15 @@ namespace alpaka
+            struct Abs<AbsStdLib, TArg>
             {
-                ALPAKA_FN_HOST auto operator()(AbsStdLib const& abs_ctx, TArg const& arg)
+                ALPAKA_FN_HOST auto operator()(AbsStdLib const& abs_ctx, ConvertedArg<TArg> const& arg)
```
So the idea for the host side is to convert `alpaka::Complex<T>` into `std::complex<T>` and just call the standard functions like we do for real numbers. Unfortunately, it didn't work as is, since typecast didn't get along with template substitution. To avoid having to specialize for `Complex` separately (which would have been needed for each math function), I've wrapped this type.

sbastrakov commented on include/alpaka/math/abs/AbsStdLib.hpp:
```diff
@@ -29,17 +30,15 @@ namespace alpaka
                     return std::abs(arg);
                 }
             };
+
```
Accidental change

sbastrakov commented on include/alpaka/math/abs/AbsUniformCudaHipBuiltIn.hpp:
```diff
@@ -83,6 +84,16 @@ namespace alpaka
                     return ::fabsf(arg);
                 }
             };
+            //! The CUDA built in abs Complex specialization.
+            template<typename T>
+            struct Abs<AbsUniformCudaHipBuiltIn, Complex<T>>
+            {
+                __device__ auto operator()(AbsUniformCudaHipBuiltIn const& abs_ctx, Complex<T> const& arg)
+                {
+                    alpaka::ignore_unused(abs_ctx);
+                    return ::sqrt(arg.real() * arg.real() + arg.imag() * arg.imag());
+                }
+            };
```
Since core CUDA doesn't have any complex type (only CUDA libraries do), we will have to just implement everything manually

sbastrakov replied:
Probably `::sqrt` is wrong for `T == float`, and I should have called alpaka's `sqrt()` instead

SimeonEhrig replied:
Do you mean with cuda complex type a class or struct (I think so)? There are a lot of complex types in CUDA, but every complex type is a alias of float2.

sbastrakov replied:
There are in CUDA libraries, not in CUDA itself as far as I understand

sbastrakov replied:
There is indeed `cuComplex` which is an alias for `float2`. Weirdly, it is not searchable in the programming guide.

SimeonEhrig replied:
Attention, there are more complex types in the CUDA libraries, e.g. `cuFFT`
```C
typedef cuComplex cufftComplex;
typedef cuDoubleComplex cufftDoubleComplex;
```

But should be all a typedef of `float2`.

bernhardmgruber commented on include/alpaka/math/Complex.hpp:
```diff
@@ -0,0 +1,333 @@
+        }
+
+        //! Stub for get the real part (does nothing, added for compatibility with std::complex)
+        constexpr ALPAKA_FN_HOST_ACC void real(T)
+        {
+        }
```
I think that is supposed to be the setter, so you are missing a `m_real = ...;`. See: https://en.cppreference.com/w/cpp/numeric/complex/real
Same for `imag()`.

sbastrakov replied:
True, I was thinking of smth else. Thanks for pointing out

bernhardmgruber commented on include/alpaka/math/Complex.hpp:
```diff
@@ -0,0 +1,333 @@
+        template<typename U>
+        ALPAKA_FN_HOST_ACC Complex& operator/=(Complex<U> const& other)
+        {
+            return *this *= Complex{
+                       static_cast<T>(other.real() / (other.real() * other.real() + other.imag() * other.imag())),
+                       static_cast<T>(-other.imag() / (other.real() * other.real() + other.imag() * other.imag()))};
```
Here is a subtlety: part of the computation is done in Us, another part in Ts. I think this needs some thinking what is right here.

sbastrakov replied:
I think it's supposed to be mixed in cppreference definition. As an alternative, I could first cast the other argument to `Complex<T>` and call the same method with matching `T`.

bernhardmgruber reviewed (commented):
I am wondering whether the complex type should be templated on the accelerator. Because then you could just use `std::complex` everywhere on the CPU and only supply a different implementation for CUDA/HIP. I think this is also what @sliwowitz does with his RNG. AFAIK there is a different Philox implementation depending on the Acc.

sbastrakov commented:
During the VC @bernhardmgruber suggested if it is woth to have `Acc` as an additional template parameter of `Complex`. Then it would be possible to partially specialize there, maybe for CUDA/HIP enabled vs. disabled. Then device-side functions would change from `Complex<T>` like now to `Complex<TAcc, T>` or `TComplex`, and generally it becomes futher away from `std::complex<T>`

sbastrakov commented:
Notes from @SimeonEhrig on the VC. In the example we need to demostrate more: how to cast between buffer of `alpaka::Complex` and `std::complex`, how to call math operations on host (either cast to `std::complex` and call, or have a kernel called on `Host` acc). I agree, will extend the example.

psychocoderHPC commented on include/alpaka/math/Complex.hpp:
```diff
@@ -0,0 +1,333 @@
+        }
+
+        //! Subtraction assignment with a real number
+        ALPAKA_FN_HOST_ACC Complex& operator-=(T const& other)
```
I think we can define the operators constexpr too?

sbastrakov replied:
I defined same as `std::complex` in cppreference for C++14. It says these operators will become `constexpr` since C++20.

psychocoderHPC replied:
Sry, I missed that this is the `operator-=` and others are defined constexpr.

BenjaminW3 commented on include/alpaka/math/ConvertedArg.hpp:
```diff
@@ -0,0 +1,55 @@
+            //!
+            //! \tparam TArg original argument type
+            template<typename TArg>
+            using ConvertedArg = typename detail::ConvertedArg<TArg>::type;
```
From the name it is not clear for the user what is converted to which type and why? When should the user specialize this trait?
Some more explicit naming might help.

sbastrakov replied:
Yes, I was struggling to find a fitting name here. It's also not for the users of alpaka, only for its developers.

sbastrakov replied:
This is no longer needed, removed in the current version.

SimeonEhrig commented:
I tested complex numbers with vikunja: https://github.com/alpaka-group/vikunja/pull/54
The API usage is nice and it works well with the CPU backend. But I have a problem with the CUDA backend. If I compile it for the CUDA backend, I get the following error:

```
/opt/spack-modules/linux-ubuntu20.04-skylake_avx512/gcc-10.2.0/cuda-11.0.2-mtxiuzc35ykbmjk4fnpmhwnjbgquvcv5/bin/nvcc -forward-unknown-to-host-compiler -ccbin=/usr/bin/c++ -DALPAKA_ACC_CPU_B_SEQ_T_SEQ_ENABLED -DALPAKA_ACC_GPU_CUDA_ENABLED -DALPAKA_BLOCK_SHARED_DYN_MEMBER_ALLOC_KIB=47 -DALPAKA_DEBUG=0 -DALPAKA_DEBUG_OFFLOAD_ASSUME_HOST -DALPAKA_OFFLOAD_MAX_BLOCK_SIZE=256 -DBOOST_ALL_NO_LIB -I../vikunja/include -I/opt/spack-modules/linux-ubuntu20.04-skylake_avx512/gcc-10.2.0/cuda-11.0.2-mtxiuzc35ykbmjk4fnpmhwnjbgquvcv5/targets/x86_64-linux/include -I../alpaka/include -isystem=/opt/spack-modules/linux-ubuntu20.04-skylake_avx512/gcc-10.2.0/boost-1.74.0-lqyr7ou7iafiwvq5fb6ipp62o6l2gqlf/include -isystem=/opt/spack-modules/linux-ubuntu20.04-skylake_avx512/gcc-10.2.0/cuda-11.0.2-mtxiuzc35ykbmjk4fnpmhwnjbgquvcv5/include --generate-code=arch=compute_52,code=[compute_52,sm_52] -Xcompiler=-fPIE --expt-extended-lambda --expt-relaxed-constexpr -Xcudafe=--display_error_number -Xcudafe=--diag_suppress=esa_on_defaulted_function_ignored -Xcompiler -pthread -std=c++17 -MD -MT CMakeFiles/complexVikunja.dir/main.cpp.o -MF CMakeFiles/complexVikunja.dir/main.cpp.o.d -x cu -c ../main.cpp -o CMakeFiles/complexVikunja.dir/main.cpp.o
../alpaka/include/alpaka/math/abs/Traits.hpp(54): error #3050: calling a __device__ function("operator()") from a __host__ __device__ function("abs") is not allowed
          detected during instantiation of "auto alpaka::math::abs(const T &, const TArg &) [with T=alpaka::AccGpuCudaRt<alpaka::DimInt<1UL>, uint64_t>, TArg=alpaka::Complex<float>]" 
../main.cpp(114): here

../alpaka/include/alpaka/math/abs/Traits.hpp(54): error #3050: calling a __device__ function("operator()") from a __host__ __device__ function("abs") is not allowed
          detected during instantiation of "auto alpaka::math::abs(const T &, const TArg &) [with T=alpaka::AccGpuCudaRt<alpaka::DimInt<1UL>, uint64_t>, TArg=alpaka::Complex<float>]" 
../main.cpp(114): here

2 errors detected in the compilation of "../main.cpp".
ninja: build stopped: subcommand failed.
```

This line is causing the error
https://github.com/alpaka-group/vikunja/blob/28c82eada54e042adc316a57c9ef172c21bc1834/example/complex/main.cpp#L117

The problem has to do with the specialization of the `acc` object. Maybe I use it wrong.

sbastrakov commented:
Thanks for reporting. I will try.

psychocoderHPC commented:
I discussed @SimeonEhrig issue with him offline, it could be he must change https://github.com/alpaka-group/vikunja/blob/28c82eada54e042adc316a57c9ef172c21bc1834/example/complex/main.cpp#L116 to

```C++
auto transform = [] ALPAKA_FN_HOST_ACC(auto const& acc, alpaka::Complex<Data> const& a) -> Data
```

to enforce the compile to evaluate the lambda on the device side only. IMO the issue is coming from the two-phase compiling of the cuda compiler.

SimeonEhrig commented:
After an additional discussion with @psychocoderHPC we found a solution. Now I use a `struct` with templated `operator()` function for the transform (I was not aware, that is already working :sweat_smile: ). Now the template injection is working correctly and the whole application too.

I uploaded the change: https://github.com/alpaka-group/vikunja/blob/77ed0ce48b33721d3d5a90705968817682a98121/example/complex/main.cpp#L19

Therefore, I'm fine with the complex numbers.

sbastrakov commented:
Just pushed a mostly finished version. Will hopefully fix the failing tests and do some polish tomorrow, also will clean up commit history.

sbastrakov commented:
@SimeonEhrig there was an issue in my original code that caused a mixup of host and host-device math. Now I believe it is fixed for everything but `arg` and `conj`, however so far i do not see what's exactly wrong with them

So perhaps your original code was fine and the error was on my side, I do not know.

SimeonEhrig commented:
> @SimeonEhrig there was an issue in my original code that caused a mixup of host and host-device math. Now I believe it is fixed for everything but `arg` and `conj`, however so far i do not see what's exactly wrong with them

Looks like, I was not affected of this bug. Nevertheless I tested you latest commit (bd35e7beb96305d3b51284204aeef9584c2c20ce) with my vikunja example and following acc's successful.
* AccCpuSerial
* AccCpuOmp2Blocks
* AccGpuCudaRt
* AccGpuHipRt

sbastrakov commented on include/alpaka/math/arg/Traits.hpp:
```diff
@@ -0,0 +1,60 @@
+            template<typename T, typename TArgument, typename TSfinae = void>
+            struct Arg
+            {
+                //!@todo why is this needed here, but not in other functions?
+                ALPAKA_NO_HOST_ACC_WARNING
```
I must be missing something very simple. But i have no idea why is there a warning generated here about host and device functions (unless I add a guard). The function works as intended (on host and device, for real, complex, and other types (the other are with ADL)), as tests show and I also checked manually. And in similar situations other functions do not generate such a warning. If anyone has an idea, it's welcome.

Btw the same is for another new math function/trait added in this PR, `conj()`

sbastrakov commented on test/unit/math/src/math.cpp:
```diff
@@ -125,12 +133,47 @@ struct TestTemplate
+        // Consider all infinite values equal, @see comment at isFinite()
+        if(!isFinite(a) && !isFinite(b))
+            return true;
+        // For the same reason use relative difference comparison with quite a large epsilon
+        auto const eps = static_cast<T>(1.0e4) * std::numeric_limits<T>::epsilon();
```
Oh damn, I forgot to change back this part

j-stephan commented:
Shifting to 0.9. This is a feature too important to put into a release on such short notice.

psychocoderHPC commented:
@sbastrakov Please do not forget to rebase before pushing updates.

j-stephan commented:
What is the state here?

psychocoderHPC commented:
@sbastrakov Could you finish this PR?

sbastrakov committed:
Add alpaka::Complex<T>, mostly a host-device version of std::complex<T>

sbastrakov committed:
Extend math traits for Complex, add new traits for arg() and conj()

sbastrakov committed:
Extend unit tests for Complex

sbastrakov committed:
Add an example demonstrating use of Complex

sbastrakov commented:
I updated the old PR state to match the current alpaka `develop`. Math tests build and run on the dev system. However I believe the previously encountered issues with accuracy threshold for unit tests on some CI configurations still remain (since i didn't do anything to fix them yet).

sbastrakov commented:
SYCL backend is not updated yet, added a checkbox.

sbastrakov committed:
Split math unit tests into multiple translation units

After adding tests for Complex<>, Visual Studio CI setup was consistently running out of heap memory.
Split the tests into separate translation units for float, double, complex float, complex double, and all ADL tests separately.
The split fixes this issue at the cost of slight code duplication.

sbastrakov commented:
CI passed, I just need to clean up history a bit, since it took some effort to make it pass.

Ready for review @j-stephan @bernhardmgruber

bernhardmgruber commented on example/complex/CMakeLists.txt:
```diff
@@ -0,0 +1,60 @@
+#
+# Copyright 2014-2022 Erik Zenker, Benjamin Worpitz, Jan Stephan, Sergei Bastrakov
```
```suggestion
# Copyright 2022 Erik Zenker, Benjamin Worpitz, Jan Stephan, Sergei Bastrakov
```

bernhardmgruber commented on include/alpaka/math/Complex.hpp:
```diff
@@ -0,0 +1,540 @@
+    //! Additionally, provides an implicit conversion to and from std::complex<T>.
+    //! All methods besides operators << and >> are host-device.
+    //! Note that it does not provide non-member functions of std::complex besides the operators.
+    //! Those are provided the same way as alpaka math functions for real numbers.
```
```suggestion
    //! It follows the layout of std::complex and so array-oriented access.
    //! The class template implements all methods and operators as std::complex<T>.
    //! Additionally, it provides an implicit conversion to and from std::complex<T>.
    //! All methods besides operators << and >> are host-device.
    //! It does not provide non-member functions of std::complex besides the operators.
    //! Those are provided the same way as alpaka math functions for real numbers.
```

bernhardmgruber commented on include/alpaka/math/Complex.hpp:
```diff
@@ -0,0 +1,540 @@
+    //! Note that it does not provide non-member functions of std::complex besides the operators.
+    //! Those are provided the same way as alpaka math functions for real numbers.
+    //!
+    //! Note that unlike most of alpaka, this is a concrete type template, and not merely a concept.
```
I would not need that comment. But you can keep it.

sbastrakov replied:
I guess it's quite obvious, but after seeing our traits for math a new user might have expected a similar generalized complex implementation. So i'll prefer to keep it.

bernhardmgruber commented on include/alpaka/math/Complex.hpp:
```diff
@@ -0,0 +1,540 @@
+    //!
+    //! Naming and order of the methods match https://en.cppreference.com/w/cpp/numeric/complex.
+    //!
+    //! @tparam T type of the real and imaginary part: float, double, or long double.
+    template<typename T>
+    class Complex
+    {
```
```suggestion
    //! @tparam T type of the real and imaginary part: float, double, or long double.
    template<typename T>
    class Complex
    {
        static_assert(std::is_floating_point_v<T>);
```
Maybe enforce that `T` is a floating point in addition to the comment.

bernhardmgruber commented on include/alpaka/math/Complex.hpp:
```diff
@@ -0,0 +1,540 @@
+            return std::complex<T>{m_real, m_imag};
+        }
+
+        //! Assignment
+        ALPAKA_FN_HOST_ACC Complex& operator=(Complex const&) = default;
+
```
```suggestion
```
Just use the compiler provided one. Or are you getting a warning if you don't provide a defaulted one?

sbastrakov replied:
I think i had a warning when using a compiler-provided one, but will check for the current code version.

bernhardmgruber replied:
Some tools warn if you provide a copy constructor but now copy assignment operator. So this might be a reason to have it.

sbastrakov replied:
Yes, i actually now faced it. So will keep as it was.

bernhardmgruber replied:
Ok!

bernhardmgruber commented on include/alpaka/math/Complex.hpp:
```diff
@@ -0,0 +1,540 @@
+        }
+
+        //! Addition assignment with a real number
+        ALPAKA_FN_HOST_ACC Complex& operator+=(T const& other)
+        {
+            m_real += other;
+            return *this;
+        }
```
Why not `constexpr` the operators as well?

sbastrakov replied:
I wanted to make them as close to `std::complex<>` as possible, and there the operators become `constexpr` in C++20.

bernhardmgruber replied:
Ok, that explains the choice. I would still make it `constexpr` because it makes the class more powerful and also forward compatible to C++20.

sbastrakov replied:
So perhaps i will just add a `@todo` to the header doxygen of the class to switch to C++20 signatures of `std::complex` when we transition

sbastrakov replied:
Discussed in mattermost, will keep C++17 interface and add a note on motivation of it.

bernhardmgruber commented on include/alpaka/math/Complex.hpp:
```diff
@@ -0,0 +1,540 @@
+        //! Real and imaginary parts, storage enables array-oriented access
+        T m_real, m_imag;
+    };
+
```
Maybe add:
```c++
#ifdef __cpp_lib_is_layout_compatible
static_assert(std::is_layout_compatible_v<Complex<float>, std::complex<float>>);
#endif
```
here or in one of the unit tests.

sbastrakov replied:
[This page](https://en.cppreference.com/w/cpp/utility/feature_test) lists  `__cpp_lib_is_layout_compatible` as C++20.

bernhardmgruber replied:
Which is perfectly fine! When the macro is not defined in C++17, then the `static_assert` is just bypassed.

sbastrakov replied:
Okay, i'll implement it

bernhardmgruber commented on include/alpaka/math/Complex.hpp:
```diff
@@ -0,0 +1,540 @@
+    //! @{
+    //!
+
+    //! Unary plus (added for compatibility with std::complex)
+    template<typename T>
+    ALPAKA_FN_HOST_ACC Complex<T> operator+(Complex<T> const& val)
+    {
+        return val;
+    }
```
Prefer to define the arithmetic operators as hidden friends inside the class. Applies to all operators here.

sbastrakov replied:
I probably would for a new class, but here mirrored [how cppreference describes `std::complex`](https://en.cppreference.com/w/cpp/numeric/complex/operator_arith2)

bernhardmgruber commented on include/alpaka/math/Complex.hpp:
```diff
@@ -0,0 +1,540 @@
+    constexpr ALPAKA_FN_HOST_ACC bool operator!=(Complex<T> const& lhs, Complex<T> const& rhs)
+    {
+        return !math::floatEqualExactNoWarning(lhs.real(), rhs.real())
+            || !math::floatEqualExactNoWarning(lhs.imag(), rhs.imag());
```
```suggestion
        return !(lhs == rhs);
```
Reuse `==` when implementing `!=`. Applies to the operator `!=` implementations as well.

bernhardmgruber commented on include/alpaka/math/Complex.hpp:
```diff
@@ -0,0 +1,540 @@
+    //! @}
+
+    //! Host-only output of a complex number
+    template<typename T, typename TChar, typename TTraits>
+    std::basic_ostream<TChar, TTraits>& operator<<(std::basic_ostream<TChar, TTraits>& os, Complex<T> const& x)
+    {
+        os << x.operator std::complex<T>();
+        return os;
+    }
```
You really want to support wide streams? :D

sbastrakov replied:
No, but i mirrored [how `std::complex` does it](https://en.cppreference.com/w/cpp/numeric/complex/operator_ltltgtgt)

bernhardmgruber replied:
It's fine! It's just really the first time I see someone use `basic_ostream` instead of `ostream`.

bernhardmgruber commented on include/alpaka/math/Complex.hpp:
```diff
@@ -0,0 +1,540 @@
+        }
+
+        //! Addition assignment with a complex number
+        template<typename U>
+        ALPAKA_FN_HOST_ACC Complex& operator+=(Complex<U> const& other)
+        {
+            m_real += static_cast<T>(other.real());
+            m_imag += static_cast<T>(other.imag());
+            return *this;
+        }
```
```suggestion
        ALPAKA_FN_HOST_ACC Complex& operator+=(Complex const& other)
        {
            m_real += other.real();
            m_imag += other.imag();
            return *this;
        }
```
For `operator+=` you support converting from complex numbers with different FP type, but e.g. `operator+` does not support his conversion. I would keep it stricter and only allow a matching complex number in arithmetic operations.

sbastrakov replied:
I re-checked, that's what `std::complex` does, only assignment-operation has mixed-type support

bernhardmgruber replied:
Ok then!

bernhardmgruber commented on include/alpaka/math/Complex.hpp:
```diff
@@ -0,0 +1,540 @@
+        constexpr ALPAKA_FN_HOST_ACC Complex(Complex const& other) = default;
+
+        //! Constructor from Complex of another type
+        template<typename U>
+        constexpr ALPAKA_FN_HOST_ACC Complex(Complex<U> const& other)
+            : m_real(static_cast<T>(other.real()))
+            , m_imag(static_cast<T>(other.imag()))
+        {
+        }
```
```suggestion
        template<typename U>
        explicit constexpr ALPAKA_FN_HOST_ACC Complex(Complex<U> const& other)
            : m_real(static_cast<T>(other.real()))
            , m_imag(static_cast<T>(other.imag()))
        {
        }
```
You may consider making conversion between complex numbers of different FP type explicit.

sbastrakov replied:
It's non-explicit [here](https://en.cppreference.com/w/cpp/numeric/complex/complex), so i did the same

bernhardmgruber replied:
Fine!

bernhardmgruber commented on test/unit/math/src/Functor.hpp:
```diff
@@ -51,11 +70,22 @@ namespace alpaka
             typename TAcc = std::nullptr_t,                                                                           \
-            typename... TArgs, /* SFINAE: Enables if called from host. */                                             \
+            typename TArg1, /* SFINAE: Enables if called from host. */                                                \
+            typename std::enable_if<std::is_same<TAcc, std::nullptr_t>::value, int>::type = 0>                        \
```
```suggestion
            std::enable_if_t<std::is_same_v<TAcc, std::nullptr_t>, int> = 0>                        \
```

SimeonEhrig commented:
@sbastrakov If you are finish with applying the suggestions, I would test the PR again with the vikunja example: https://github.com/alpaka-group/vikunja/pull/54

sbastrakov commented:
@bernhardmgruber thanks for your review! As i answered in a few points, the choices were made simply to mirror interface of `std::complex<T>`. The idea is, that it's as smooth as possible to transition from `std::complex<T>` to `alpaka::Complex<T>` when porting a code base to alpaka. So personally i would also do what you suggest in some places, but think here it's not worth to deviate from the `std::` implementation unless there is a good reason.

I resolved the points i implemented in my local branch, and kept open and answered the points I think should not be implemented.

sbastrakov committed:
Implement review suggestions

bernhardmgruber commented:
@sbastrakov the PR is fine as is. I only suggested a few enhancements to improve the implementation. I would really like to see the `constexpr` and layout comaptible check. But if you don't want to add it, so be it.

bernhardmgruber commented on test/unit/math/src/mathComplexFloat.cpp:
```diff
@@ -34,3 +35,17 @@ TEMPLATE_LIST_TEST_CASE("mathOpsComplexFloat", "[math] [operator]", TestAccFunct
     testTemplate.template operator()<alpaka::Complex<float>>();
 }
+
+#ifdef __has_include
+#    if __has_include(<version>)
+#        include <version>
+#    endif
+#endif
```
`__cpp_lib_is_layout_compatible` is also defined inside `<type_traits>`, so you can just include that header instead, which you will need anyway for `std::is_layout_compatible_v`.

bernhardmgruber commented on test/unit/math/src/mathComplexFloat.cpp:
```diff
@@ -0,0 +1,51 @@
+TEMPLATE_LIST_TEST_CASE("mathOpsComplex", "[layout]", TestAccs)
+{
+    // Check both float and double here to avoid duplicating the wrapper code
+    REQUIRE(std::is_layout_compatible_v<alpaka::Complex<float>, std::complex<float>>);
+    REQUIRE(std::is_layout_compatible_v<alpaka::Complex<double>, std::complex<double>>);
```
```suggestion
    STATIC_REQUIRE(std::is_layout_compatible_v<alpaka::Complex<float>, std::complex<float>>);
    STATIC_REQUIRE(std::is_layout_compatible_v<alpaka::Complex<double>, std::complex<double>>);
```
make it a compile-time check.

SimeonEhrig commented:
> @sbastrakov If you are finish with applying the suggestions, I would test the PR again with the vikunja example: [alpaka-group/vikunja#54](https://github.com/alpaka-group/vikunja/pull/54)

I tested the example with Serial CPU, OpenMP Grid-Block, CUDA and HIP backkend. Looks good.

sbastrakov commented:
Damn, now 1 test that have passed yesterday fails. Even tho i'm sure i didn't make any changes that should have affected it.

I now loosened the bounds a little bit, hopefully will pass now. Also implemented @bernhardmgruber 's latest review suggestion.

sbastrakov commented:
@bernhardmgruber @j-stephan ready to be merged
